src/dune/action_exec.ml

```diff
@@ -238,21 +308,34 @@ and redirect_in t ~ectx ~eenv inputs fn =
     | Stdin -> in_
   in
   exec t ~ectx ~eenv:{ eenv with stdin_from }
-  >>| fun () -> Process.Io.release in_
+  >>| fun result ->
+  Process.Io.release in_;
+  result
 
 and exec_list ts ~ectx ~eenv =
   match ts with
-  | [] -> Fiber.return ()
+  | [] -> Fiber.return Done
   | [ t ] -> exec t ~ectx ~eenv
-  | t :: rest ->
-    let* () =
+  | t :: rest -> (
+    let* done_or_deps =
       let stdout_to = Process.Io.multi_use eenv.stdout_to in
       let stderr_to = Process.Io.multi_use eenv.stderr_to in
       let stdin_from = Process.Io.multi_use eenv.stdin_from in
       exec t ~ectx ~eenv:{ eenv with stdout_to; stderr_to; stdin_from }
     in
-    exec_list rest ~ectx ~eenv
+    match done_or_deps with
+    | Need_more_deps _ as need -> Fiber.return need
```

Comment:
Actually, it might be worth defining and using the `('a, needs_more_deps) Result.t Fiber.t` monad. That might reduce the diff in this file significantly, in particular `exec_list` might not need any change at all then.

Reply:
See staronj/dune#1, I'm not sure if it's worth it but I might just overcomplicated it.

Reply:
OK, I guess it's not worth it. It's not really noticeably better on any metric, it seems.